pkg/common/idutil/safety.go

```diff
@@ -176,18 +144,130 @@ func IDFromProto(id *types.SPIFFEID) (spiffeid.ID, error) {
 // FormatPath formats a path string. The function ensures a leading slash is
 // present.
 func FormatPath(format string, args ...interface{}) string {
+	// TODO: when the safety valve is removed, this function should:
+	// 1. not ensure the leading slash.
+	// 2. validate that the produced path is correct
 	return ensureLeadingSlash(fmt.Sprintf(format, args...))
 }
 
 // JoinPathSegments escapes path segments and joins them together. The
 // function also ensures a leading slash is present.
 func JoinPathSegments(segments ...string) string {
+	// TODO: when the safety valve is removed, this function should:
+	// 1. not ensure the leading slash.
+	// 2. validate that the produced path is correct
 	return ensureLeadingSlash(strings.Join(segments, "/"))
 }
 
+// TrustDomainFromString parses a trust domain from a string.
+func TrustDomainFromString(s string) (spiffeid.TrustDomain, error) {
+	td, err := spiffeid.TrustDomainFromString(s)
+	if err != nil {
+		return spiffeid.TrustDomain{}, err
+	}
+	if !allowUnsafeIDs() {
+		if err := validateTrustDomain(td.String()); err != nil {
+			return spiffeid.TrustDomain{}, err
+		}
+
+		// spiffeid.TrustDomainFromString currently "normalizes" the trust
+		// domain portion by lowercasing. We don't want to mask an "invalid"
+		// trust domain by this normalization, so we do this prefix check here
+		// to detect if it happened. The input string should prefix match
+		// either trust domain name or trust domain ID otherwise some
+		// normalization occurred.
+		if !strings.HasPrefix(s, td.String()) && !strings.HasPrefix(s, td.IDString()) {
+			return spiffeid.TrustDomain{}, errors.New("trust domain can only contain [a-z0-9.-]")
+		}
+	}
+	return td, nil
+}
+
 func ensureLeadingSlash(p string) string {
 	if p != "" && p[0] != '/' {
 		p = "/" + p
 	}
 	return p
 }
+
+func urlParse(id string) (*url.URL, error) {
+	// Detect an errant scheme beforehand since url.Parse will lowercase the
+	// scheme automatically.
+	if !strings.HasPrefix(id, "spiffe://") {
+		return nil, errors.New("scheme must be 'spiffe'")
+	}
+	return url.Parse(id)
+}
+
+func validateComponents(td, path string) error {
+	if err := validateTrustDomain(td); err != nil {
+		return err
+	}
+	return validatePath(path)
+}
+
+func validateTrustDomain(td string) error {
+	for i := 0; i < len(td); i++ {
+		if !isValidTrustDomainChar(td[i]) {
+			return errors.New("trust domain can only contain [a-z0-9.-]")
+		}
+	}
+	return nil
+}
+
+func validatePath(path string) error {
+	segment := 0
+	i := 0
+	for ; i < len(path); i++ {
+		c := path[i]
+		if c == '/' {
+			switch path[segment:i] {
+			case "/":
+				return errors.New("path cannot contain empty segments")
+			case "/.", "/..":
+				return errors.New("path cannot contain dot segments")
+			}
+			segment = i
+			continue
+		}
+		if !isValidPathSegmentChar(c) {
+			return errors.New("path can only contain [a-zA-Z0-9.-_]")
```

Comment:
```suggestion
			return errors.New("path can only contain letters, numbers, dots, dashes, and underscores")
```

While most SPIRE operators probably know regex, it's best to not expect them to. Same comment for `validateTrustDomain` and `TrustDomainFromString`

Reply:
:+1: